python/pyarrow/tests/test_table.py

```diff
@@ -335,12 +348,34 @@ def test_table_remove_column_empty():
     table = pa.Table.from_arrays(data, names=['a'])
 
     t2 = table.remove_column(0)
+    t2._validate()
     assert len(t2) == len(table)
 
     t3 = t2.add_column(0, table[0])
+    t3._validate()
     assert t3.equals(table)
 
 
+def test_table_flatten():
+    ty1 = pa.struct([pa.field('x', pa.int16()),
+                     pa.field('y', pa.float32())])
+    ty2 = pa.struct([pa.field('nest', ty1)])
+    a = pa.array([(1, 2.5), (3, 4.5)], type=ty1)
+    b = pa.array([((11, 12.5),), ((13, 14.5),)], type=ty2)
+    c = pa.array([False, True], type=pa.bool_())
+
+    table = pa.Table.from_arrays([a, b, c], names=['a', 'b', 'c'])
+    t2 = table.flatten()
+    t2._validate()
+    expected = pa.Table.from_arrays([
+        pa.array([1, 3], type=pa.int16()),
+        pa.array([2.5, 4.5], type=pa.float32()),
+        pa.array([(11, 12.5), (13, 14.5)], type=ty1),
+        c],
+        names=['a.x', 'a.y', 'b.nest', 'c'])
```

Comment:
We should have some tests to acknowledge that we have to deal with non-variable-name-like-charactesrs at some point.

Reply:
I'm not sure we should deal with them ourselves? I *think* users can reimplement the table-flattening functionality themselves using `array.flatten` (if they want another naming scheme), though I'm not sure how easy that is with the currently exposed APIs.